nemo/collections/tts/data/dataset.py

```diff
@@ -683,6 +712,26 @@ def __getitem__(self, index):
         if SpeakerID in self.sup_data_types_set:
             speaker_id = torch.tensor(sample["speaker_id"]).long()
 
+        reference_audio, reference_audio_length = None, None
+        if ReferenceAudio in self.sup_data_types_set:
+            if self.speaker_to_index_map is not None:
+                reference_index = self.sample_reference_index(index, n=1)[0]
+                reference_audio = self.featurizer.process(
+                    self.data[reference_index]["audio_filepath"],
```

Comment:
I can understand the rationale for wanting to sample the speaker embeddings and reference audios from the non-training sample, but it is pretty unorthodox. Having training examples sample from other records in the manifest is going to be complicated to maintain unless the data processor is designed specifically for the use case.

In comparison, using the embeddings from the training utterance and stopping overfiting by using a bottleneck like VAE/VQ is a lot more standard and scalable. But might not work well on smaller datasets without lots of tuning.

I am curious what other's thoughts are.

Reply:
What kind of complications do you think we will have in maintaining this random sampling code? I agree with your suggestion of using VAE/VQ as a projection layer on top of the reference/GT audio/embedding and that can be incorporated into the SpeakerEncoder class. But here this might be useful for the current use-case as well as other use-cases where we need contrastive losses and need to do negative sampling.

Reply:
It might not be too bad. Maybe we can discuss how to do it when the new data loader is in PR.

Having worked a lot with contrastive losses and negative sampling before, they tend to be manageable when you train on 1 host but can become problematic when the dataset is large enough that you have to shard it and each host only sees a relatively small subset of data. For example to optimize this code, you would want a shard data by speakers (eg. host one gets speakers 0-3, host two 4-9, etc). But that is far in the future.